File: client/src/modules/nexotype/schemas/asset/therapeutic-asset.schemas.ts
import { z } from 'zod';

// ==========================================
// Enums
// ==========================================

export const AssetTypeEnum = z.enum(['small_molecule', 'biologic', 'therapeutic_peptide', 'oligonucleotide']);

// ==========================================
// Options (for Select dropdowns)
// ==========================================

/** TherapeuticAsset.asset_type */
export const ASSET_TYPE_OPTIONS = [
  { label: 'Small Molecule', value: 'small_molecule' },
  { label: 'Biologic', value: 'biologic' },
  { label: 'Therapeutic Peptide', value: 'therapeutic_peptide' },
  { label: 'Oligonucleotide', value: 'oligonucleotide' },
] as const;

// ==========================================
// Full Schema (GET responses)
// ==========================================

export const TherapeuticAssetSchema = z.object({
  id: z.number(),
  uid: z.string().min(1).max(100),
  name: z.string().min(1).max(255),
  project_code: z.string().max(50).nullable().optional(),
  asset_type: AssetTypeEnum,
  created_at: z.string(),
  updated_at: z.string().nullable(),
});

// ==========================================
// Input Schemas (POST/PUT)
// ==========================================

export const CreateTherapeuticAssetSchema = z.object({
  uid: z.string().min(1).max(100),
  name: z.string().min(1).max(255),
  project_code: z.string().max(50).nullable().optional(),
  asset_type: AssetTypeEnum,
});

export const UpdateTherapeuticAssetSchema = z.object({
  uid: z.string().min(1).max(100).optional(),
  name: z.string().min(1).max(255).optional(),
  project_code: z.string().max(50).nullable().optional(),
  asset_type: AssetTypeEnum.optional(),
});

// ==========================================
// Types
// ==========================================

export type TherapeuticAsset = z.infer<typeof TherapeuticAssetSchema>;
export type CreateTherapeuticAsset = z.infer<typeof CreateTherapeuticAssetSchema>;
export type UpdateTherapeuticAsset = z.infer<typeof UpdateTherapeuticAssetSchema>;
export type AssetType = z.infer<typeof AssetTypeEnum>;

// ==========================================
// Label Helpers
// ==========================================

/** Human-readable labels for asset types */
export const ASSET_TYPE_LABELS: Record<AssetType, string> = {
  small_molecule: 'Small Molecule',
  biologic: 'Biologic',
  therapeutic_peptide: 'Therapeutic Peptide',
  oligonucleotide: 'Oligonucleotide',
};

/** Get human-readable label for an asset type */
export const getAssetTypeLabel = (type: string): string => {
  return ASSET_TYPE_LABELS[type as AssetType] || type;
};

// ==========================================
// Response Types
// ==========================================

export type TherapeuticAssetResponse = {
  success: boolean;
  data?: TherapeuticAsset;
  error?: string;
};

export type TherapeuticAssetsResponse = {
  success: boolean;
  data?: TherapeuticAsset[];
  count?: number;
  error?: string;
};

File: server/apps/nexotype/schemas/asset/therapeutic_asset_schemas.py
from pydantic import BaseModel, ConfigDict
from datetime import datetime

# ==========================================
# Request Schemas
# ==========================================

class TherapeuticAssetCreate(BaseModel):
    """Schema for creating a new therapeutic asset"""
    uid: str
    name: str
    project_code: str | None = None
    asset_type: str


class TherapeuticAssetUpdate(BaseModel):
    """Schema for updating a therapeutic asset"""
    uid: str = None
    name: str = None
    project_code: str | None = None
    asset_type: str = None


# ==========================================
# Response Schemas
# ==========================================

class TherapeuticAssetDetail(BaseModel):
    """Schema for therapeutic asset details"""
    id: int
    uid: str
    name: str
    project_code: str | None = None
    asset_type: str
    created_at: datetime
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class TherapeuticAssetListResponse(BaseModel):
    """Response schema for listing therapeutic assets"""
    success: bool
    data: list[TherapeuticAssetDetail] | None = None
    count: int = 0
    error: str | None = None


class TherapeuticAssetResponse(BaseModel):
    """Response schema for single therapeutic asset operations"""
    success: bool
    data: TherapeuticAssetDetail | None = None
    error: str | None = None


class MessageResponse(BaseModel):
    """Simple message response"""
    success: bool
    message: str | None = None
    error: str | None = None
